### PR TITLE
Compiler optimizations for new classes

### PR DESCRIPTION
**Pull request by zfields**
PR provides compiler optimizations by labeling the functions as `const`.
Also replaces #326 by addressing issue #324 

Only optimizes the internal functionality, and does not affect the external API.

### CODE REVIEW

**Comment by zfields** [comment]
@soundanalogous Did you have any thoughts or concerns about this PR?

**Comment by soundanalogous** [comment]
I haven't had enough time to look into it. Will look at it tomorrow. I generally do most of my work on Firmata on weekends and just check in from time to time during the week.

**Comment by zfields** [comment]
My apologies for rushing you, I'm just excited and iterating quickly. I'll dial it down a notch. :smiley:

File: Firmata.h [hunk 1]
```diff
@@ -111,7 +111,7 @@ class FirmataClass
 
     /* private methods ------------------------------ */
     void strobeBlinkPin(byte pin, int count, int onInterval, int offInterval);
-    friend void FirmataMarshaller::sendValueAsTwo7bitBytes(uint16_t value);
+    friend void FirmataMarshaller::sendValueAsTwo7bitBytes(uint16_t value) const;
```
Comment (soundanalogous) [inline Firmata.h]: What is the compiler optimization of this use of `const`. I was not familiar with this technique and just looked into it. From what I understand it informs the compiler that no class members within the scope of this method can be altered. I'm just not sure what the advantage of this declaration is.

Comment (zfields) [inline Firmata.h]: From the C++ Standard:

> **9.3.2  The this pointer**
> ...
> If the member function is declared `const`, the type of this is `const X*`, if the member
> function is declared `volatile`, the type of this is `volatile X*`, and if the member function
> is declared `const volatile`, the type of this is `const volatile X*`. *[Note: thus in a `const`
> member function, the object for which the function is called is accessed through a `const`
> access path. — end note]*

There is no guarantee, but the compiler may able to optimize around constants.

Here is an excellent [discussion on StackOverflow](http://stackoverflow.com/questions/27466642/what-kind-of-optimization-does-const-offer-in-c-c-if-any).

Here a well written [blog by Herb Sutter](http://www.gotw.ca/gotw/081.htm)

**Comment by soundanalogous** [comment]
Thanks for the links. I've tested this and there are no errors so merging.

**Pull request merged**